### PR TITLE
Master fw small improvements

### PR DESCRIPTION
**Pull request by moebio**
makes several improvements (replaces lots of for(i=0; this[i]!=null…) and fixes some issues (kMeans)

### CODE REVIEW

**Comment by Jeff-Clark** [comment]
+1 looks good
- But note that there is a functional change here. The old-style loops for(i=0; this[i]!=null…) would stop on null elements. The new method will loop and continue past nulls processing all the remaining items. I don't believe we normally intentionally store nulls in lists/tables etc but it is possible it occurs somewhere.

- Also it would be good to add some tests for some of the more complex operators like NumberTableOperators.kMeans.

**Comment by moebio** [comment]
@Jeff-Clark indeed, there's a behavior change, and as you say we don't expect nulls in XList, and we could expect nulls in List… in which case I think is better then to completely accept the possibility of null and not cut processes whenever a null appear.

I've been making these changes for the last ~3 weeks, and everything seems to continue working well (including kMeans). The replacement proved to be robust. And efficiency improved.

Only place I left list[i]!=null as criteria for continue looping is when the list is actually transformed (typically a splice). Preferred to let it that way for the efficiency gain is lower (one has to calculate the length on each change on the list), but specially because of being afraid of breaking something.

File: src/dataTypes/lists/List.js [hunk 1]
```diff
@@ -1252,7 +1275,8 @@ List.prototype.getWithoutElements = function(list) {
  */
 List.prototype.getFilteredByFunction = function(func) {
   var newList = instantiateWithSameType(this);
-  for(var i = 0; this[i] != null; i++) {
+  var l = this.length;
```
Comment (protozoo) [inline src/dataTypes/lists/List.js]: @moebio  Would declaring (not setting) this variable out of the first loop make any difference?

File: src/dataTypes/lists/Table.js [hunk 2]
```diff
@@ -36,7 +36,9 @@ Table.prototype.constructor = Table;
 function Table() {
   var args = [];
   var i;
-  for(i = 0; i < arguments.length; i++) {
+  var nArgs = arguments.length;
```
Comment (protozoo) [inline src/dataTypes/lists/Table.js]: @moebio Would declaring (not setting) this variable out of the first loop make any difference?

File: src/dataTypes/lists/Table.js [hunk 3]
```diff
@@ -136,7 +140,9 @@ Table.prototype.getListLength = function(index) {
  */
 Table.prototype.getLengths = function() {
   var lengths = new NumberList();
-  for(var i = 0; this[i] != null; i++) {
+  var l = this.length;
```
Comment (protozoo) [inline src/dataTypes/lists/Table.js]: @moebio Would declaring (not setting) this variable out of the first loop make any difference?

File: src/dataTypes/lists/Table.js [hunk 4]
```diff
@@ -113,7 +115,9 @@ Table.prototype.applyFunction = function(func) {
 Table.prototype.getRow = function(index) {
   var list = new List();
   var i;
-  for(i = 0; i < this.length; i++) {
+  var l = this.length;
```
Comment (protozoo) [inline src/dataTypes/lists/Table.js]: @moebio Would declaring (not setting) this variable out of the first loop make any difference?

**Comment by protozoo** [comment]
+1

File: src/dataTypes/lists/Table.js (hunk 4 above)
Comment (moebio) [inline src/dataTypes/lists/Table.js]: @protozoo  it does a big difference, with for(i = 0; i < this.length; i++) you're calculating the length of the array on each loop

File: src/dataTypes/lists/List.js [hunk 5]
```diff
@@ -414,6 +414,7 @@ List.prototype.getSubListByType = function(type) {
  * tags:filter
  */
 List.prototype.getSubListByIndexes = function() { //TODO: merge with getSubList
+  console.log("\n\n\n\n List.prototype.getSubListByIndexes | this: ", this);
```
Comment (tafsiri) [inline src/dataTypes/lists/List.js]: did you want to leave this in?

**Pull request merged**